### PR TITLE
4 - ByteArray module

### PR DESCRIPTION
**Pull request by TAdev0**

### CODE REVIEW

File: corelib/src/byte_array.cairo [hunk 1]
```diff
@@ -1,3 +1,47 @@
+//! A dynamic, growable array of bytes where bytes are stored in multiples of 31 bytes,
+//! with a `felt252` pending word used for bytes that don't complete a full `bytes31` word.
+//!
```
Comment (enitrat) [inline corelib/src/byte_array.cairo:3]:
```suggestion
//! A dynamic, growable array of bytes where bytes are stored in multiples of 31 bytes,
//! with a `felt252` pending word used for bytes that don't complete a full `bytes31` word.
//!
//! `ByteArray` is Cairo's `String` type, and can be created with string literals: `"hello, world!"`
```

File: corelib/src/byte_array.cairo [hunk 2]
```diff
@@ -1,3 +1,47 @@
...
+//! ```
+//!
+//! The `ByteArray` type also supports operations like addition, which can be used for
+//! concatenation:
+//!
+//! ```
+//! let mut ba: ByteArray = "";
+//! let mut other_ba: ByteArray = "";
+//! let concatenated = ba + other_ba;
+//! ```
```
Comment (enitrat) [inline corelib/src/byte_array.cairo:43]:
```suggestion
//! # Examples
//!
//! There are multiple ways to create a new `ByteArray`:
//! - From a string litteral:
//! ```
//! let s = "Hello";
//!```
//! - Using the `format!` macro:
//!```
//! let max_tps:u16 = 850;
//! let s = format!("Starknet's max TPS is: {}", max_tps);
//! ```
//! You can append bytes to an existing `ByteArray`  with [`ByteArrayTrait::append_byte`]:
//!
//! ```
//! let mut ba: ByteArray = "";
//! ba.append_byte(0x41); // Appending a single byte 'A'
//! ```
//!
//! You can create a new `ByteArray` from an existing one by concatenating with
//! `+`:
//!
//! ```
//! let s = "Hello";
//!
//! let message = s + " world!";
//! ```
//! Indexing operations are available on the `ByteArray` type as well:
//!
//! ```
//! let mut ba: ByteArray = "ABC";
//! let first_bytes = ba[0]
//! assert!(first_byte == 0x41);
//! ```
```

Changed some stuff to reflect real-world usage better

File: corelib/src/byte_array.cairo [hunk 3]
```diff
@@ -9,48 +53,60 @@ use crate::cmp::min;
...
 const BYTES_IN_BYTES31_MINUS_ONE: usize = BYTES_IN_BYTES31 - 1;
 
 // TODO(yuval): don't allow creation of invalid ByteArray?
+/// An array of `bytes31` defined as a structure containing 3 fields.
 #[derive(Drop, Clone, PartialEq, Serde, Default)]
 pub struct ByteArray {
-    // Full "words" of 31 bytes each. The first byte of each word in the byte array
-    // is the most significant byte in the word.
+    /// `data` field is an array of full "words" of 31 bytes each.
+    /// The first byte of each word in the byte array is the most significant byte in the word.
```
Comment (enitrat) [inline corelib/src/byte_array.cairo:75]:
```suggestion
    /// An array of full "words" of 31 bytes each.
    /// The first byte of each word in the byte array is the most significant byte in the word.
```

avoid formulations like "x field", "pending_word field" when you're documenting that specific field directly

File: corelib/src/byte_array.cairo [hunk 4]
```diff
@@ -9,48 +53,60 @@ use crate::cmp::min;
...
     pub(crate) data: Array<bytes31>,
-    // This felt252 actually represents a bytes31, with < 31 bytes.
-    // It is represented as a felt252 to improve performance of building the byte array.
-    // The number of bytes in here is specified in `pending_word_len`.
-    // The first byte is the most significant byte among the `pending_word_len` bytes in the word.
+    /// `pending_word` field is a `felt252` that actually represents a `bytes31`, with < 31 bytes.
+    /// It is represented as a `felt252` to improve performance of building the byte array.
+    /// The first byte is the most significant byte among the `pending_word_len` bytes in the word.
```
Comment (enitrat) [inline corelib/src/byte_array.cairo:79]:
```suggestion
    /// A `felt252` that actually represents a `bytes31`, with less than 31 bytes.
    /// It is represented as a `felt252` to improve performance of building the byte array.
    /// The first byte is the most significant byte among the `pending_word_len` bytes in the word.
```

File: corelib/src/byte_array.cairo [hunk 5]
```diff
@@ -9,48 +53,60 @@ use crate::cmp::min;
...
     pub(crate) pending_word: felt252,
-    // Should be in range [0, 30].
+    /// `pending_word_len` represents the number of bytes in `pending_word`.
```
Comment (enitrat) [inline corelib/src/byte_array.cairo:81]:
```suggestion
    /// The number of bytes in `pending_word`.
```

File: corelib/src/byte_array.cairo [hunk 6]
```diff
@@ -9,48 +53,60 @@ use crate::cmp::min;
...
+    /// Its value should be in the range [0, 30].
     pub(crate) pending_word_len: usize,
 }
 
 pub(crate) impl ByteArrayStringLiteral of crate::string::StringLiteral<ByteArray>;
 
+/// Basic trait for the `ByteArray` type.
```
Comment (enitrat) [inline corelib/src/byte_array.cairo:88]:
```suggestion
/// Functions associated with the `ByteArray` type.
```

also in line with how we present Associated Functions in the cairo book 👯

File: corelib/src/byte_array.cairo [hunk 7]
```diff
@@ -141,14 +205,31 @@ pub impl ByteArrayImpl of ByteArrayTrait {
         self.append_word(*other.pending_word, *other.pending_word_len);
     }
 
-    /// Concatenates two byte arrays and returns the result.
+    /// Concatenates two `ByteArray` and returns the result.
+    ///
+    /// # Examples
```
Comment (enitrat) [inline corelib/src/byte_array.cairo:210]:
```suggestion
    /// Concatenates two `ByteArray` and returns the result.
    ///
    /// The content of `left` is cloned in a new memory segment.
    /// # Examples
```

File: corelib/src/byte_array.cairo [hunk 8]
```diff
@@ -170,12 +251,30 @@ pub impl ByteArrayImpl of ByteArrayTrait {
         self.pending_word_len = 0;
     }
 
+    /// Returns the length of the `ByteArray` as a `usize` value.
```
Comment (enitrat) [inline corelib/src/byte_array.cairo:254]:
```suggestion
    /// Returns the length of the `ByteArray`.
```

**Review by enitrat — COMMENTED** [review]
good, some suggestions

**Comment by enitrat** [comment]
@TAdev0 unsure if you've seen these comments

File: corelib/src/byte_array.cairo (hunk 1 above)
Comment (TAdev0) [reply corelib/src/byte_array.cairo]: done

**Comment by TAdev0** [comment]
@enitrat  added your suggestion, thanks!

**Comment by TAdev0** [comment]
@orizi  made some modifications

**Pull request closed without merging**